maro/rl/training/algorithms/base/ac_ppo_base.py

```diff
@@ -202,18 +202,43 @@ def preprocess_batch(self, batch: TransitionBatch) -> TransitionBatch:
         # Preprocess advantages
         states = ndarray_to_tensor(batch.states, device=self._device)  # s
         actions = ndarray_to_tensor(batch.actions, device=self._device)  # a
+        terminals = ndarray_to_tensor(batch.terminals, device=self._device)
+        next_states = ndarray_to_tensor(batch.next_states, device=self._device)
         if self._is_discrete_action:
             actions = actions.long()
 
         with torch.no_grad():
             self._v_critic_net.eval()
             self._policy.eval()
             values = self._v_critic_net.v_values(states).detach().cpu().numpy()
-            values = np.concatenate([values, np.zeros(1)])
-            rewards = np.concatenate([batch.rewards, np.zeros(1)])
-            deltas = rewards[:-1] + self._reward_discount * values[1:] - values[:-1]  # r + gamma * v(s') - v(s)
-            batch.returns = discount_cumsum(rewards, self._reward_discount)[:-1]
-            batch.advantages = discount_cumsum(deltas, self._reward_discount * self._lam)
+
+            batch.returns = np.zeros(batch.size, dtype=np.float32)
```

Comment:
Is it required to be a full trajectory?
How to guarantee it?

Reply:
It is not required to be a full trajectory. If the last trajectory in the experiences is partial, we only need to use `critic(next_state)` as the `last_value` rather than 0 to calculate return.